2023.10.2 compile issue

### The problem

```
|-- Improv @ 1.2.3
Compiling .pioenvs/esp32-bluetooth-proxy-0d77f4/src/esphome/components/api/api_connection.o
Compiling .pioenvs/esp32-bluetooth-proxy-0d77f4/src/esphome/components/api/api_frame_helper.o
Compiling .pioenvs/esp32-bluetooth-proxy-0d77f4/src/esphome/components/api/api_pb2.o
Compiling .pioenvs/esp32-bluetooth-proxy-0d77f4/src/esphome/components/api/api_pb2_service.o
In file included from src/esphome/components/network/util.h:4,
                 from src/esphome/components/api/api_connection.cpp:4:
src/esphome/components/network/ip_address.h: In constructor 'esphome::network::IPAddress::IPAddress(ip4_addr_t*)':
src/esphome/components/network/ip_address.h:40:14: error: 'ip_addr_t' {aka 'struct ip4_addr'} has no member named 'type'
     ip_addr_.type = IPADDR_TYPE_V4;
              ^~~~
Compiling .pioenvs/esp32-bluetooth-proxy-0d77f4/src/esphome/components/api/api_server.o
*** [.pioenvs/esp32-bluetooth-proxy-0d77f4/src/esphome/components/api/api_connection.o] Error 1
In file included from src/esphome/components/network/util.h:4,
                 from src/esphome/components/api/api_server.cpp:4:
src/esphome/components/network/ip_address.h: In constructor 'esphome::network::IPAddress::IPAddress(ip4_addr_t*)':
src/esphome/components/network/ip_address.h:40:14: error: 'ip_addr_t' {aka 'struct ip4_addr'} has no member named 'type'
     ip_addr_.type = IPADDR_TYPE_V4;
              ^~~~
*** [.pioenvs/esp32-bluetooth-proxy-0d77f4/src/esphome/components/api/api_server.o] Error 1
========================= [FAILED] Took 118.55 seconds =========================
```

### Which version of ESPHome has the issue?

2023.10.2

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2023.10.5

### What platform are you using?

ESP32-IDF

### Board

AZ Delivery ESP32S

### Component causing the issue

bluetooth proxy

### Example YAML snippet

```yaml
trying to update my bluetooth proxy in esphome. Initial install was with Esphome Web
```


### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

the same here. LAN module wt32-eth01 and wifi module esp32

```
Compiling .pioenvs/wt32-eth01-bt-proxy-macaddr/src/esphome/components/api/api_pb2_service.o
In file included from src/esphome/components/network/util.h:4,
                 from src/esphome/components/api/api_connection.cpp:4:
src/esphome/components/network/ip_address.h: In constructor 'esphome::network::IPAddress::IPAddress(ip4_addr_t*)':
src/esphome/components/network/ip_address.h:40:14: error: 'ip_addr_t' {aka 'struct ip4_addr'} has no member named 'type'
     ip_addr_.type = IPADDR_TYPE_V4;
              ^~~~
Compiling .pioenvs/wt32-eth01-bt-proxy-macaddr/src/esphome/components/api/api_server.o
*** [.pioenvs/wt32-eth01-bt-proxy-macaddr/src/esphome/components/api/api_connection.o] Error 1
In file included from src/esphome/components/network/util.h:4,
                 from src/esphome/components/api/api_server.cpp:4:
src/esphome/components/network/ip_address.h: In constructor 'esphome::network::IPAddress::IPAddress(ip4_addr_t*)':
src/esphome/components/network/ip_address.h:40:14: error: 'ip_addr_t' {aka 'struct ip4_addr'} has no member named 'type'
     ip_addr_.type = IPADDR_TYPE_V4;
              ^~~~
*** [.pioenvs/wt32-eth01-bt-proxy-macaddr/src/esphome/components/api/api_server.o] Error 1
========================= [FAILED] Took 28.92 seconds =========================
```

Same here on Atom Echo.

```
Compiling .pioenvs/atom-echo-1/src/esphome/components/esp32/preferences.o
In file included from src/esphome/components/network/util.h:4,
                 from src/esphome/components/api/api_connection.cpp:4:
src/esphome/components/network/ip_address.h: In constructor 'esphome::network::IPAddress::IPAddress(ip4_addr_t*)':
src/esphome/components/network/ip_address.h:40:14: error: 'ip_addr_t' {aka 'struct ip4_addr'} has no member named 'type'
     ip_addr_.type = IPADDR_TYPE_V4;
              ^~~~
Compiling .pioenvs/atom-echo-1/src/esphome/components/esp32_rmt_led_strip/led_strip.o
Compiling .pioenvs/atom-echo-1/src/esphome/components/esp_adf/esp_adf.o
Compiling .pioenvs/atom-echo-1/src/esphome/components/gpio/binary_sensor/gpio_binary_sensor.o
Compiling .pioenvs/atom-echo-1/src/esphome/components/homeassistant/time/homeassistant_time.o
In file included from src/esphome/components/network/util.h:4,
                 from src/esphome/components/api/api_server.cpp:4:
src/esphome/components/network/ip_address.h: In constructor 'esphome::network::IPAddress::IPAddress(ip4_addr_t*)':
src/esphome/components/network/ip_address.h:40:14: error: 'ip_addr_t' {aka 'struct ip4_addr'} has no member named 'type'
     ip_addr_.type = IPADDR_TYPE_V4;
              ^~~~
Compiling .pioenvs/atom-echo-1/src/esphome/components/i2s_audio/i2s_audio.o
Compiling .pioenvs/atom-echo-1/src/esphome/components/i2s_audio/microphone/i2s_audio_microphone.o
In file included from src/esphome/components/wifi/wifi_component.h:7,
                 from src/esphome/components/captive_portal/captive_portal.cpp:4:
src/esphome/components/network/ip_address.h: In constructor 'esphome::network::IPAddress::IPAddress(ip4_addr_t*)':
src/esphome/components/network/ip_address.h:40:14: error: 'ip_addr_t' {aka 'struct ip4_addr'} has no member named 'type'
     ip_addr_.type = IPADDR_TYPE_V4;
              ^~~~
Compiling .pioenvs/atom-echo-1/src/esphome/components/i2s_audio/speaker/i2s_audio_speaker.o
Compiling .pioenvs/atom-echo-1/src/esphome/components/improv_base/improv_base.o
*** [.pioenvs/atom-echo-1/src/esphome/components/api/api_connection.o] Error 1
*** [.pioenvs/atom-echo-1/src/esphome/components/captive_portal/captive_portal.o] Error 1
*** [.pioenvs/atom-echo-1/src/esphome/components/api/api_server.o] Error 1
In file included from src/esphome/components/network/util.h:4,
                 from src/esphome/components/improv_base/improv_base.cpp:3:
src/esphome/components/network/ip_address.h: In constructor 'esphome::network::IPAddress::IPAddress(ip4_addr_t*)':
src/esphome/components/network/ip_address.h:40:14: error: 'ip_addr_t' {aka 'struct ip4_addr'} has no member named 'type'
     ip_addr_.type = IPADDR_TYPE_V4;
              ^~~~
*** [.pioenvs/atom-echo-1/src/esphome/components/improv_base/improv_base.o] Error 1
========================== [FAILED] Took 6.83 seconds ==========================
```

--- a/esphome/components/network/ip_address.h
+++ b/esphome/components/network/ip_address.h
@@ -36,7 +36,7 @@ struct IPAddress {
   IPAddress(const std::string &in_address) { ipaddr_aton(in_address.c_str(), &ip_addr_); }
   IPAddress(ip4_addr_t *other_ip) {
     memcpy((void *) &ip_addr_, (void *) other_ip, sizeof(ip4_addr_t));
-#if USE_ESP32
+#if defined(USE_ARDUINO) && defined(USE_ESP32)
     ip_addr_.type = IPADDR_TYPE_V4;
 #endif
   }

--- a/esphome/components/network/ip_address.h
+++ b/esphome/components/network/ip_address.h
@@ -36,7 +36,7 @@ struct IPAddress {
   IPAddress(const std::string &in_address) { ipaddr_aton(in_address.c_str(), &ip_addr_); }
   IPAddress(ip4_addr_t *other_ip) {
     memcpy((void *) &ip_addr_, (void *) other_ip, sizeof(ip4_addr_t));
-#if defined(USE_ARDUINO) && defined(USE_ESP32)
+#if USE_ESP32 && LWIP_IPV6
     ip_addr_.type = IPADDR_TYPE_V4;
 #endif
   }

